products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -41,54 +41,94 @@ public class BuildDBKeysResult {
     
     private lazy var _count: Int = Int(llb_database_result_keys_get_count(result))
     
-    /// Get the key at the given index
-    public func getKey(at index: KeyID) -> BuildKey {
-        BuildKey.construct(key: llb_database_result_keys_get_key_at_index(self.result, index))
-    }
-    
     deinit {
         llb_database_destroy_result_keys(result)
     }
 }
 
 extension BuildDBKeysResult: Collection {
-    public typealias Index = KeyID
+    public typealias Index = Int
     public typealias Element = BuildKey
     
     public var startIndex: Index {
         return 0
     }
     
     public var endIndex: Index {
-        return UInt64(self._count) + startIndex
+        return self._count + startIndex
     }
     
     public subscript(index: Index) -> Iterator.Element {
         guard (startIndex..<endIndex).contains(index) else {
             fatalError("Index \(index) is out of bounds (\(startIndex)..<\(endIndex))")
         }
-        return getKey(at: index)
+        return BuildKey.construct(key: llb_database_result_keys_get_key_at_index(self.result, Int32(index)))
     }
     
-    public func index(after i: KeyID) -> KeyID {
+    public func index(after i: Index) -> Index {
         return i + 1
     }
 }
 
 extension BuildDBKeysResult: CustomReflectable {
     public var customMirror: Mirror {
-        let keys = (startIndex..<endIndex).map { getKey(at: $0) }
+        let keys = (startIndex..<endIndex).map { self[$0] }
         return Mirror(BuildDBKeysResult.self, unlabeledChildren: keys, displayStyle: .collection)
     }
 }
 
 extension BuildDBKeysResult: CustomStringConvertible {
     public var description: String {
-        let keyDescriptions = (startIndex..<endIndex).map { "\(getKey(at: $0))" }
+        let keyDescriptions = (startIndex..<endIndex).map { "\(self[$0]))" }
         return "[\(keyDescriptions.joined(separator: ", "))]"
     }
 }
 
+/// Defines the result of a built task
+public class RuleResult {
+    /// Reference to the C API result instance (for desctruction)
+    private let result: BuildDBResult
+    /// The value of the result
+    public let value: BuildValue
+    /// Signature of the node that generated the result
+    public let signature: UInt64
+    /// The build iteration this result was computed at
+    public let computedAt: UInt64
+    /// The build iteration this result was built at
+    public let builtAt: UInt64
+    /// The start of the command as a duration since a reference time
+    public let start: Double
+    /// The duration since a reference time of when the command finished computing
+    public let end: Double
+    /// The duration in seconds the result needed to finish
+    public var duration: Double {
+        max(end - start, 0)
```

Comment:
It feels like `end >= start` should be an assert and we should just return the raw diff. Given that we are using a monotonic clock, deviation from that expectation should not be the case.  And if we are, it is potentially unexpected that duration doesn't match `end - start`

Reply:
Or in other words, I don't think the bindings layer is the right layer for such logic to exist.

Reply:
Yep, that's true